### PR TITLE
Solution to the test

### PR DESCRIPTION
**Pull request by franleplant**
Ok, here's my take on it. 

Vercel deploy: https://yelp-resto-explorer.vercel.app/

Github Repo: https://github.com/franleplant/frontend-test

## Highlights

- use `next.js` and deploy it to `vercel` in very simple steps
- use `typescript` effectively and efficiently by letting `graphql-codegen` to generate ts types out of the gql schema
- use `emotion` for styles and theming and sharing variables
- everything is automatically formatted by `prettier`
- use `graphql` and `apollo` for effective and efficient data fetching
- use `react-testing-library` for good unit testing of react components
- use `jest` as a amazing testing framework with good support for `mocks` out of the box
- use `storybook` as a component explorer / development environment
- use css `grid layout` for responsive grid views
- use `React Portals` to create a very simple but powerful implementation of a `Tooltip` see src/components/Tooltip
- good overall project structure that I have used to scale up to a million lines of code
- all the data accessing is abstracted in the `dal` (data access layer) layer and relies heavily on the gql codegen for types, queries and hooks
- cool use of gql `Fragments` in order to be able to reference them as Typescript types in the rest of the app
- use of `css` variables as a efficient way of passing dynamic runtime props to emotion styled components
- `CI` setup via github actions
- `CD` setup via vercel (mostly automagically)

A lot of these things such as CI/CD, gql-codegen, typescript, emotion theming, jest, storybook are
part of a really good boilerplate for any app and that I've used in the past (and in the present), a lot
of care has gone to those areas in order to represent what I value and prioritize.

## Things I would improve but I de-prioritized due to time constraints

- improve Apollo isomorphic features to allow certain data to be fetched in the sever
- improve A11y support via eslint analyzers (tried a few but were confused by emotion)
- add more unit tests
- add user level testing? react testing library is pretty good and high level though
- mobile view
- animation and cosmetics
- review the theme / styling and try to abstract more building blocks

### CODE REVIEW

File: pages/index.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  );
+}
+
+export const Header = styled.header`
+  ${containerStyles}
+
+  // trick for extra specifity without important
+  && > * + * {
+    margin-top: 24px;
+  }
```
Comment (alex-sf) [inline pages/index.tsx:37]: This is a cool little trick but why do we need extra specificity in the first place? What is it fighting against? and how could we structure things so that it wouldnt need to make this type of assertion in the first place?

File: src/components/SelectPrice/index.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  { label: "All", value: "1,2,3" },
+  { label: "$", value: "1" },
+  { label: "$$", value: "2" },
+  { label: "$$$", value: "3" },
+];
+
+export interface IProps {
+  value: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  onChange: (newPrice: string) => any;
```
Comment (alex-sf) [inline src/components/SelectPrice/index.tsx:14]: I noticed a lot of the `onChange` handlers return an `any` type. Can you talk about the choice to use that instead of something like `void`?

**Review by alex-sf — COMMENTED** [review]
Thanks for getting this in, it looks great 🎉  I left a few comments for you on the PR. There's nothing you need to change in the code in response to these, just respond to these comments as if it was a code review.

File: .env [hunk 3]
```diff
@@ -0,0 +1,3 @@
+YELP_GRAPHQL_URI=https://api.yelp.com/v3/graphql
+YELP_CLIENT_ID=TODO
+YELP_API_KEY=TODO
```
[the remark this answers is not in the document]
Comment (franleplant) [reply .env:3]: I think this is a pretty good one but you could do a couple of major things

- add documentation to the README about it
- have a configuration service that provides this (java/spring has something like this baked in)
- use typescript.

For the last one you could the obvious of overriding the type of `process.env` (via interface merging) although
when the type you want to overwrite is very lax I haven't had good experiences (i.e. like trying to type express' `req.locals`), so I think that a much better option would be to do this:

```typescript

const MY_ENV = {
   VAR1: process.env.VAR1,
   VAR2: process.env.VAR2,
   ...
   VARN: process.env.VARN,
}

export type IEnv = typeof MY_ENV

export default MY_ENV
```

With this setup you get the best of all worlds:

- you statically type your env
- your env type is completely connected to the run time type
- there is a sync between the inferred static type and the real run time type.

You could even make this more strict by having an accessor that throws when a var is missing:

```typescript

export function getEnv(key, required=true): string {
  const value = process.env[key]
  if (!value && required) {
   throw new Error(`missing env ${key}`)
  }
  return value
}
```

And so with these two things you have static and runtime guarantees about your env but also
really helpful documentation (because the first time the MY_ENV file runs it will try to read
all the necessary variables and report an error if some are missing).

All this can be combined with the `.env` solution which I think is pretty nice.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Use Node.js ${{ matrix.node-version }}
+      uses: actions/setup-node@v1
+      with:
+        node-version: ${{ matrix.node-version }}
+    - run: yarn
+    - run: yarn test
+    - run: yarn build
```
[the remark this answers is not in the document]
Comment (franleplant) [reply .github/workflows/ci.yml:28]: Im not 100% sure but I would expect to be able to configure github to prevent merging PRs that don't pass the CI process (most alternatives do have that capability). 

Yeap, this is an integral part of the CD pipeline (although in this case the CD pipeline goes through vercel so further configuration on that side should be taken but lets assume for a moment that the CD pipeline is also part of this github action), if you and your team have good typescript coverage and good test coverage then you can easily (at least in the lower environments) make this the first required step of a CD pipeline, if all these tests and checks pass then you can feel comfortable automatically deploying it to lower envs. 

This is how most CI/CD pipelines work: install deps, run checks, build, deploy, I work with this type of setup and for example my personal blog has this setup as well https://github.com/franleplant/nosleepjavascript-blog

You could even replace the "deploy" part for a "publish to npm" script and this can also apply to libraries.

TLDR: if a PR does not have passing tests and builds then it means that it is not working and something needs to be fixed and thus it should not be merged into master/main/trunk

File: package.json [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    "jest": "^26.6.3",
+    "lint-staged": ">=10",
+    "prettier": "^2.2.1",
+    "prettier-eslint": "^12.0.0",
+    "ts-node": "^9.1.1",
+    "typescript": "^4.1.3"
+  },
+  "husky": {
+    "hooks": {
+      "pre-commit": "lint-staged"
```
[the remark this answers is not in the document]
Comment (franleplant) [reply package.json:72]: I think this ends up being a personal preference, it is common to run `yarn test` on `prepush` (doing `git push origin mybranch` locally), the real test though is on the CI server because the CI server should resemble your PROD (or lower envs) servers.

This pre-commit hook _only_ does autoformatting which is a really quality of life since you basically can forget about prettier on 99% of the cases.

I've seen `yarn test` and other stuff being run on `commit` but I think that is too awkward for several reasons like for example `not every commit must be a working commit` and experienced developers can always bypass those checks via `git commit -m "whatever" --no-verify`, so again, the real thing happens on the CI server, everything else should be quality of life for developers, that is why I personally only do prettier after every commit and not much more.

File: pages/_app.tsx [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import { ThemeProvider } from "@emotion/react";
+import { getClient } from "../src/dal/apollo";
+import { theme } from "../src/theme";
+import GlobalStyles from "../src/components/GlobalStyles";
+
+export interface IProps {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Component: React.ComponentType<any>;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  pageProps: any;
```
[the remark this answers is not in the document]
Comment (franleplant) [reply pages/_app.tsx:12]: Assuming you are talking about _my_ implementation I wouldn't say there is, I just didn't use that typing because I didn't spend the extra time to look for it but I would definitively use it. I think that the main reason they use a class based component is that they are declaring a couple of static methods but you could do something similar with functional components by doing basically the same but with a bit of more complicated typing i.e. `NextApp.getInitialProps = ....`, assuming NextApp is a function then you get the same, in fact the old way of doing classes was sort of like that and a lot of transpiler use that old version when representing classes.

If you are talking about class component vs functional components then... I would say that today there isn't a lot of hard differences between one another except the interface they achieve, the static methods is a nice difference, but for the most part I think that functional components with hooks is far superior in all respects, I think that hooks are both weird and an amazing way of abstracting, reusing and composing side effects (I have a course in the making about that).

There also seems to be some usage of old context and extras as noted here https://github.com/vercel/next.js/issues/7515

File: pages/index.tsx (hunk 1 above)
Comment (franleplant) [reply pages/index.tsx:37]: I have completely forgotten why I had to do this in the first place, perhaps it was something needed but now it isn't, it was basically fighting with other class that was defining `marging=-top` for one of the Header's children and had a higher specificity and broke it.

grid and flex `gap` css attribute can be used to replaced this and I've actually used it elsewhere.

In general I try to avoid using margin in lower level components (such as buttons, inputs, etc) and let the higher level (in the tree) components that deal with margin to set it up to avoid this fighting.

Sorry for the vague response.

File: src/components/SelectPrice/index.tsx (hunk 2 above)
Comment (franleplant) [reply src/components/SelectPrice/index.tsx:14]: Yeap, great question: so in my experience handlers (mostly custom ones) can return `void` or `Promise<void>`, so instead of typing them as `void | Promise<void>` I tend to use `any` because we really do not care about what type they return, if I were to be super strict I would probably use `void | Promise<void>` and if I were to care about the return type then I would definitively type that appropriately.

I am open to be convinced otherwise though.

Why would a handler return `Promise<void>` ? async functions:

```typescript
async function onChange(newValue: string ): Promise<void> {
  await doSomethingAsync(newValue)
}

//...
<MyInput onChange={onChange} />
```

Comment (julian-sf) [reply src/components/SelectPrice/index.tsx:14]: I believe Alex's question is less about the return type of the particular functions and more about the usage of the `any` type. `any` is a fully-permissive construct, that allows any type through.

Using `any` essentially "disables" TypeScript and forces consumers to be either dangerous themselves with the result or use judicious type guards to return back into type safety. It's generally considered an anti-pattern and will lead to more complex or dangerous code down the road.

To answer the question of "Why would a handler return Promise<void>?" that's simple, maybe you have a login or logout function that sets some JWT or cookie. And you want to allow consumers to "wait" for the operation to finish before continuing.

```typescript
await login(credentials)
```

This is one reason that `Promise<void>` can be useful.

But a more pertinent point is: you don't NEED to type for `Promise<void> | void` to accept whatever you want. Just defining `onChange` as `(newPrice: string) => void` does what you actually want, in a type safe manner.

```typescript
export interface IProps { onChange: (newPrice: string) => void }

function syncExample() {}
async function asyncExample() {}
function numberSyncExample() { return 1 }
async function numberAsyncExample() { return 1 }

const synced: IProps = { onChange: syncExample }
const asynced: IProps = { onChange: asyncExample }
const numberSynced: IProps = { onChange: numberSyncExample }
const numberAsynced: IProps = { onChange: numberAsyncExample }
```

File: src/dal/categories/index.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    if (!categories) {
+      continue;
+    }
+
+    for (const cat of categories) {
+      if (!cat) {
+        continue;
+      }
+
+      totalCategories[cat.alias!] = cat;
```
Comment (julian-sf) [inline src/dal/categories/index.ts:41]: **question:** what happens if `cat.alias` is not defined?

How can you guard against this?

**Review by julian-sf — COMMENTED** [review]
## Thanks for your submission!

I have reviewed your code and posted a series of comments throughout. Most of these are questions to gauge your technical abilities. They are not an indictment or even a sign of code issues. We generally just want to probe around and see how you think and approach problems.

Thank you for being patient with us during the review process. I look forward to hearing back from you!

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [14.x, 15.x]
+        # See supported Node.js release schedule at https://nodejs.org/en/about/releases/
```
[the remark this answers is not in the document]
Comment (franleplant) [reply .github/workflows/ci.yml:18]: I mostly kept the default template that github actions output, I agree that having a broad version coverage makes more sense for libraries although in this particular case for apps it might make sense to test the latest LTS version (currently 14) which should be the main version supported across the system but also test the next version just to check forward compatibility and to hopefully spot any future problems that the app might have migrating to the next LTS (16).

In general for an app I would use the latest LTS across the board (CI, docker images that run in all envs even in production).

And in this case I should probably configure the CI process to only warn when the tests in node 15 fail.

I also inherited this from the Rust community where they specifically ask folks building code to run their code in the unstable channels (beta and nightly) just to give the Rust team more information about potential regressions, so in a sense you could use this setup as a way of providing early feedback to the Node.js community too.

File: README.md [hunk 9]
```diff
@@ -1,117 +1,61 @@
...
+```
 
-> What if I have a question?
+## TODOs
 
-Just create a new issue in this repo and we will respond and get back to you quickly.
+- improve Apollo isomorphic features to allow certain data to be fetched in the sever
+- improve A11y support via eslint analyzers (tried a few but were confused by emotion)
+- add more unit tests
+- add user level testing? react testing library is pretty good and high level though
```
[the remark this answers is not in the document]
Comment (franleplant) [reply README.md:61]: In the most traditional sense I would have unit tests (similar to what I did), user/browser level testing (using Selenium, protractor, Webdriver.io or Cypress) (also called e2e tests), and finally a QA level tests.

In the different organizations I have been the line between what QA does and what the dev team does has always been blurry, but I had made a bunch of things with selenium/webdriver and I have been spying on Webdriver.io and Cypress and they look pretty fine. The main reasons I didn't attempt to include them is because of time.

Additionally, thanks to the combination of Jest and testing-library I think you get a pretty good higher level test coverage with these tools plus, with testing-library, you kinda get forced to do better a11y which is always a plus (because they recommend relying on aria attributes such as roles, etc).

I think that Jest makes testing fun again because of the mock functionality that works really well in practice and lets you test complex stuff in a simple and fun way.

I made a post about it https://nosleepjavascript.com/why-i-love-jest/

Additionally there is what in traditional literature called Integration tests that I would say fall between unit tests and user level tests but in practice the previously described techniques are what I've used with a great degree of success.

Perhaps in the case of Rust you can a closer thing to integration tests https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests because they are all hooked as an outside consumer.

File: fetchSchema.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * A simple solution for the fetching of the yelp api
+ * schema for later typescript codegen. I tried using
+ * grqpahl-codegen's custom loaders but it didn't work
+ * so I resorted to out of band fetching.
+ *
+ * This will store the schema in schema.json and that
+ * will be read by the code gen process, it is actually
+ * probably not needed to run this very often
```
[the remark this answers is not in the document]
Comment (franleplant) [reply fetchSchema.ts:18]: I would say that once every PR? but ideally this should automated, creating like automated PRs periodically with potential updates to the schema.

I don't think I get the last question, the main purpose of this whole setup is that you derive typescript types (and functions) from the real graphql schema and then you simply just use that in the rest of your app, so if anything changes in your schema that breaks your app code then the expected scenario is that a type error will be thrown by typescript and then you would simply make the changes.
Does that answser your question?

File: fetchSchema.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  GraphQLSchema,
+  printIntrospectionSchema,
+  getIntrospectionQuery,
+} from "graphql";
+import got from "got";
+
+/**
+ * A simple solution for the fetching of the yelp api
+ * schema for later typescript codegen. I tried using
+ * grqpahl-codegen's custom loaders but it didn't work
```
[the remark this answers is not in the document]
Comment (franleplant) [reply fetchSchema.ts:13]: The plain [URL](https://graphql-code-generator.com/docs/getting-started/schema-field#url) option doesn't work because the graphql api requires authentication, I tried using headers but I couldn't make the header value be an interpolation of `process.env.GRAPHQL_URL`.

Then I tried [running a js file](https://graphql-code-generator.com/docs/getting-started/schema-field#javascript-export) but it didn't work, IIRC because of not supporting promises

And finally I tried a [custom loader](https://graphql-code-generator.com/docs/getting-started/schema-field#custom-schema-loader) but I couldn't also make promises work.

I looked for issues without any success and I was pretty far behind in time so I went for the "easy" solution of fetching the schema out-of-band and then simply telling the codegen to read it as a json, which ended up working and was super easy to setup (downloading the graphql schema was a bit undocumented so I looked a how libraries did it and I was good to go).

With more times I would inspect the code of graphql-code, report a bug / stack overflow and even create a PR with a fix if needed.

File: pages/restaurants/[id].tsx [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  const businessId = router.query.id;
+  // TODO fetch this data from the server
+  const { data, loading, error } = useGetRestaurantQuery({
+    variables: { id: businessId as string },
+  });
+  const resto = data?.business;
+
+  const rating = resto?.rating || 0;
+  const score = (rating / 5) * 100;
+  const cat = resto?.categories?.[0]?.title;
```
[the remark this answers is not in the document]
Comment (franleplant) [reply pages/restaurants/[id].tsx:23]: That's a good question, it seems that there is no way that `cat` should be undefined, it is just the way that graphql codegen generates the types, but I agree that this could use a better worst case handling, the only question is what to do? is it an error to have an undefined `cat` inside a restaraunt? should we allow the app to fail completely? or should we default to `unknown`? or to `""`, I really don't know but I agree that this is mostly rough.

File: src/components/Button/index.tsx [hunk 13]
```diff
@@ -0,0 +1,71 @@
+import styled from "@emotion/styled";
+
+// TODO probably this should expand
+// the native button props
+export interface IProps {
```
[the remark this answers is not in the document]
Comment (franleplant) [reply src/components/Button/index.tsx:5]: I think this is the most clear and simple way of naming the type of props for the component you are sitting right now. They are both colocated in the same file, at that stage there is no confusion what so ever and the name `IProps` is as clear as water.

There are two concerns remaining:

- how to reference this potentially duplicated name in other files/components/modules? In practice I don't encounter this problem too often but in the few cases I have then the solution is super easy `import {IProps as IButtonProps} from '../components/Button'`
- will this make type errors harder to read? In my experience they don't because the actual name of the interface you are failing to comply ends up being a tiny detail in the error that typescript produces:

Real world example with some details omitted

```bash
$ tsc --noEmit
Consumer.tsx:xxx:y - error TS2741: Property 'users' is missing in type '{ onReconnect: () => Promise<void>; }' but required in type 'IProps'.

148         <Component
             ~~~~~~~~~~~~~~~~~

  Component.tsx:6:3
    6   users: Array<IUsers>
        ~~~~~~~~
    'users' is declared here.


Found 1 error.

error Command failed with exit code 2.

```

The alternative is to always name the components and the props in a related ways i.e. `Car and ICarProps` but this doesn't scale well for more complicatedly named components `ActiveUserDashboard and IActiveUserDashboardProps`, it is redundant and scanning the file for the type of the props for the component is harder.

People relying too much on full text search of codebases might find this problematic, and if the team has enough of those then I would be fine switching to other approach.

BTW: all of my answers assume that the structure of your codebase is mostly one component per file, if that's not the case then this structure doesn't work, or needs tweaks.

**Comment by franleplant** [comment]
@julian-sf  some of the issues were done in such a way that I cannot reply to them, so this is where I would reply them:

## [I prefix](https://github.com/Superformula/frontend-test/pull/40#discussion_r571051162)

I agree that it is a matter of taste so I would be fine without it but in practice it has worked really well for me.
Why do I use it?

- I like making obvious things obvious (so there is no question that something prefix with `I` will be an interface)
- In some cases I got obvious conflicting names i.e. `class Car` and `interface ICar`, by using the prefix they can both coexist without aliasing.
- And finally, and this applies mostly to typescript libraries, interfaces and types are exported differently by typescript and so they usually get lost in the `import * as MyModule from './MyModule'` supposing that `MyModule` exports interfaces and runtime values (this only applies to libraries exporting types, I cant find documentation but I am sure this is something that happens, that is why most libraries rely on namespaces for exporting types more easily). So basically making them clearly distinct i.e by prefixing them with `I` helps you manage this case because you can either attach all interfaces to a namespace or import them explicitly from the library's entry point.

Im not entirely sold on the mentioned SO question, in particular point 3 and 4, I think that it is really common to have a Class and an associated interface, because it so common in javascript/typescript to go from plain objects into classes and vice versa (think of fetching data from the server and sending it back). So if I would to do what points 3 and 4 will do I would have to create a `PlainOldCar` or `CarObject` or `SerializableCar` or `CarAttributes` etc.



## [css trick](https://github.com/Superformula/frontend-test/pull/40#discussion_r571053836)

I either don't understand what you are saying or don't agree with what you are saying. I am specifically styling the children from the parent which is a much more natural way of defining layouts. What is the alternative? to setup the needed margin in each children? And to accept it as props? It would be a more complex way of expressing this same thing. This is a very simple way of applying layout, it you were to move this component elsewhere the layout would break, which is the same thing that would happen with all layout systems out there that I know of (bootstrap's, ant's, material's, flex, css grid)

This is actually almost 100% equivalent to using `display:flex` and `gap:24px`, it is the parent, who controls the layout, dictating how the children should be laid out in the screen. This can also be equivalent to a more traditional grid system (like bootstrap's) and it's gutter / gap settings or even the css grid layout using also the `gap` attribute.

I learned this trick here https://piccalil.li/blog/cube-css

## [any on handlers](https://github.com/Superformula/frontend-test/pull/40#discussion_r571077768)

Hm, I would have gambled money that a function type that returns `void` complains about actually returning stuff, perhaps that was the behavior before? Good suggestion though.

In any case please don't take this as me advocating to overly use `any` for all function signatures or everywhere in your code, I understand completely its implications and I was only suggesting that for this case (with my previous assumptions now disproved by you) it was justified, but TLDR of course, `any` should be used to a minimum everywhere. You can also use `unknown` which is better.

File: src/dal/categories/index.ts (hunk 7 above)
Comment (franleplant) [reply src/dal/categories/index.ts:41]: It would create a `totalCategories.undefined` attribute.

This is one of those situations where there is really no right answer, the first reason this is typed as possibly missing is because of the way graphql codegen (and graphql itself) work in this particular case but given the fact that `cat.alias` seems to be a user facing `id` for the category I would expect it to always be there but I agree that something more robust could be done. in this case I would `skip and report`:

```typescript
for (const cat of categories) {
      // if cat or cat.alias are falsy (i.e. undefined) SKIP and REPORT
      if (!cat || !cat.alias) {
        logger.warn('encountered a bad category', cat)
        continue;
      }

      totalCategories[cat.alias] = cat;
    }
```

You could do other stuff like trying to coerce an alias out of the `title` but that would hurt the current way we filter by category (using the alias).

**Comment by franleplant** [comment]
done replying @julian-sf

File: fetchSchema.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (julian-sf) [reply fetchSchema.ts:18]: Sorry, the last question wasn't super clear. What I meant was "how would you keep your deployed code up to date (if no features are currently going out)?" but you answered this well. I think your ideas of automated periodic PRs would work well in that scenario.

Ideally, service changes won't happen to existing applications without some notification & preparation, right? 😆

File: src/components/Button/index.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (julian-sf) [reply src/components/Button/index.tsx:5]: Thanks for the well thought out answer!
